v1.2.0 release notes (#5596)
spiceai/spiceai: docs/release_notes/v1.2.0.md

- Add additional test scenarios for benchmarks (#5582) by @phillipleblanc in https://github.com/spiceai/spiceai/pull/5582
- fix: Update the tpch benchmark snapshots for: test/spicepods/tpch/sf1/accelerated/databricks[delta_lake]-duckdb[file].yaml (#5590) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5590
- fix: Update the tpch benchmark snapshots for: test/spicepods/tpch/sf1/accelerated/mysql-duckdb[file].yaml (#5591) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5591
- Fix Snowflake data connector rows ordering (#5599) by @sgrebnov in https://github.com/spiceai/spiceai/pull/5599
- fix: Update benchmark snapshots (#5595) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5595
- fix: Update the tpch benchmark snapshots for: test/spicepods/tpch/sf1/accelerated/databricks[delta_lake]-arrow.yaml (#5594) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5594
- fix: Update benchmark snapshots (#5589) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5589

> - fix: Update benchmark snapshots (#5583) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5583

Reviewer: ```suggestion
- fix: Update benchmark snapshots (#5583) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5583

**Full Changelog**: [v1.1.2...v1.2.0](https://github.com/spiceai/spiceai/compare/v1.1.2...v1.2.0)
```

# Spice v1.2.0 (Apr 28, 2025)

> Spice v1.2.0 is a major release which includes upgrades to DataFusion v45, Arrow 54, and DuckDB v1.2, including improvements in query peformance, and introduces parameterized query support across SQL and HTTP APIs and accelerated views. Numerous bug fixes and dependency upgrades further improve stability and performance.

Reviewer 2: I don't know if we want to avoid the specific term major here as to not conflict with the meaning from semantic versioning.

Example `spicepod.yaml` configuration:

```yaml
models:
  - from: databricks:llama-3_2_1_1b_instruct
    name: llama-instruct
    params:
      databricks_endpoint: dbc-46470731-42e5.cloud.databricks.com
      databricks_token: ${ secrets:SPICE_DATABRICKS_TOKEN }
```

- **`spice chat` CLI improvements**: The `spice chat` CLI command now supports an optional `--temperature` parameter and allows for one-shot chat requests using `spice chat <message>`.

- **Additional type support**: Added support for Postgres JSON type and DuckDB Dictionary type.

- **Additional Improvements**:
  - Added specific image tags for selecting specific memory allocators tailored to specific use-cases, including `jemalloc`, `sysalloc`, and `snmalloc`.
  - Improved error handling and logging for chat and model operations.

## Contributors

- [@peasee](https://github.com/peasee)
- [@ewgenius](https://github.com/ewgenius)
- [@Jeadie](https://github.com/Jeadie)
- [@Sevenannn](https://github.com/Sevenannn)
- [@lukekim](https://github.com/lukekim)
- [@phillipleblanc](https://github.com/phillipleblanc)
- [@sgrebnov](https://github.com/sgrebnov)
- [@kczimm](https://github.com/kczimm)

## Breaking Changes

> ### DuckDB v1.2.0

Reviewer 3: The DuckDB version is not consistent in release note, it's 1.2.2 on line 21

- The `random` function now uses a larger state, so fixed seeds will produce different values than previous DuckDB versions.
- `map['entry']` now returns a value instead of a list. Use `map_extract` to return a list, and `map_extract_value` (alias for `[]`) to return a value.
- The indexing for `list_reduce` is fixed: the index now points to the last lambda parameter and starts from 1.

No other breaking changes in this release.

## Cookbook Updates

New recipes for:

> - Using Paramerized Queries from Python over ADBC.

Reviewer 2: ```suggestion
- Using Parameterized Queries from Python over ADBC.
```

> - Using Parameertized Queries from Java.

Reviewer 2: ```suggestion
- Using Parameterized Queries from Java.
```

- Add additional test scenarios for benchmarks (#5582) by @phillipleblanc in https://github.com/spiceai/spiceai/pull/5582
- fix: Update the tpch benchmark snapshots for: test/spicepods/tpch/sf1/accelerated/databricks[delta_lake]-duckdb[file].yaml (#5590) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5590
- fix: Update the tpch benchmark snapshots for: test/spicepods/tpch/sf1/accelerated/mysql-duckdb[file].yaml (#5591) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5591
- Fix Snowflake data connector rows ordering (#5599) by @sgrebnov in https://github.com/spiceai/spiceai/pull/5599
- fix: Update benchmark snapshots (#5595) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5595
- fix: Update the tpch benchmark snapshots for: test/spicepods/tpch/sf1/accelerated/databricks[delta_lake]-arrow.yaml (#5594) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5594
- fix: Update benchmark snapshots (#5589) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5589

> - fix: Update benchmark snapshots (#5583) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5583

Reviewer: ```suggestion
- fix: Update benchmark snapshots (#5583) by @app/github-actions in https://github.com/spiceai/spiceai/pull/5583

**Full Changelog**: [v1.1.2...v1.2.0](https://github.com/spiceai/spiceai/compare/v1.1.2...v1.2.0)
```

      description: |
        Distributes requests between 'foo' and 'bar' models in a round-robin fashion.
      models:
        - from: foo
        - from: bar
    - name: fallback
      description: |
        Tries 'bar' first, then 'foo', then 'baz' if earlier models fail.
      models:
        - from: foo
          order: 2
        - from: bar
          order: 1
        - from: baz
          order: 3
  ```

- **Databricks Model Support**: Databricks models can now be used with `from: databricks:model_name`.

  Example `spicepod.yaml`:

  ```yaml
  models:
    - from: databricks:llama-3_2_1_1b_instruct
      name: llama-instruct
      params:
        databricks_endpoint: dbc-46470731-42e5.cloud.databricks.com
        databricks_token: ${ secrets:SPICE_DATABRICKS_TOKEN }
  ```

See the [Databricks model documentation](https://spiceai.org/docs/components/models/filesystem).

- **`spice chat` CLI Improvements**: The `spice chat` command now supports an optional `--temperature` parameter. A one-shot chat can also be sent with `spice chat <message>`.

- **More Type Support**: Added support for Postgres JSON type and DuckDB Dictionary type.

- **Other Improvements**:

>   - New image tags let you pick memory allocators for different use-cases: `jemalloc`, `sysalloc`, and `snmalloc`.

Reviewer 4: ```suggestion
  - New image tags let you pick memory allocators for different use-cases: `jemalloc`, `sysalloc`, and `mimalloc`.
```

- **Better Quality 📋**: DataFusion now runs over 5 million SQL tests per push using the SQLite sqllogictest suite. There are new checks for logical plan correctness and more thorough pre-release testing.

- **New SQL Functions ✨**: Added `show functions`, `to_local_time`, `regexp_count`, `map_extract`, `array_distance`, `array_any_value`, `greatest`, `least`, and `arrays_overlap`.

See the [DataFusion 45.0.0 release notes](https://datafusion.apache.org/blog/2025/02/20/datafusion-45.0.0/) for details.

Spice.ai upgrades to the latest minus one DataFusion release to ensure adequate testing and stability. The next upgrade to DataFusion v46 is planned for Spice v1.3.0 in May.

## What's New in v1.2.0

- **DuckDB v1.2.2 Upgrade**: DuckDB has been upgraded to v1.2.2 (was v1.1.3). This brings better performance and optimizations. See the [DuckDB v1.2.0 release notes](https://duckdb.org/2025/02/05/announcing-duckdb-120.html).

> - **Parameterized Queries**: Parameterized queries are now supported with the Flight SQL API and HTTP API. Both `$1` and `?` placeholders are supported. Logical plans for SQL statements are cached for faster repeated queries.

Reviewer 2: ```suggestion
- **Parameterized Queries**: Parameterized queries are now supported with the Flight SQL API and HTTP API. Positional and named arguments via `$1` and `:param` syntax are supported, respectively. Logical plans for SQL statements are cached for faster repeated queries.
```
